apm-agent-plugins/apm-logging-plugin/apm-tomcat-logging-plugin/src/main/java/co/elastic/apm/agent/tomcatlogging/reformatting/TomcatReformattingHelper.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.tomcatlogging.reformatting;
+
+import co.elastic.apm.agent.jul.reformatting.AbstractJulEcsReformattingHelper;
+import co.elastic.apm.agent.loginstr.reformatting.Utils;
+import org.apache.juli.FileHandler;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.logging.Handler;
+
+public class TomcatReformattingHelper extends AbstractJulEcsReformattingHelper {
+
+    private static final ThreadLocal<String> currentDirectory = new ThreadLocal<>();
+    private static final ThreadLocal<String> currentPrefix = new ThreadLocal<>();
+    private static final ThreadLocal<String> currentSuffix = new ThreadLocal<>();
+
+    TomcatReformattingHelper() {
+    }
+
+    public boolean onAppendEnter(FileHandler fileHandler, String directory, String prefix, String suffix) {
+        try {
+            currentDirectory.set(directory);
+            currentPrefix.set(prefix);
+            currentSuffix.set(suffix);
```

Comment:
[personal preference] I don't know why, but using ThreadLocals to just pass parameters along a code-path we 100% control feels a bit hacky to me.
An alternative would be to extend the `AbstractEcsReformattingHelper` methods `onAppendEnter` and `getShadeFilePatternAndCreateDir` with a generic `C context` parameter which gets passed through. However, because this Helper is the only case where it is needed so far, this might not really be worth it.

Reply:
I agree with you, this is a bit more hack-ish that it should, the challenge here is that:
- we can only read the class internal fields from the advice code, which is the same for regular JUL and Tomcat Juli.
- we can avoid allocating "context" objects in the case where the log events are not actually written (this depends on configuration).

So, while it definitely does not meet most code sanity standards, I would say this is fine for this case.
Alternatively, we could also use context objects as you suggested, but that would require to pool them to avoid allocation.

Reply:
Fully agree that it's fine as is, especially because the use of the `ThreadLocals` is just locally here and therefore doesn't hinder understandability :)